src/mono/mono/metadata/class.c

```diff
@@ -4100,7 +4100,7 @@ mono_class_is_assignable_from_general (MonoClass *klass, MonoClass *oklass, gboo
 			return;
 		}
 
-		if (m_class_is_array_special_interface (klass) && m_class_get_rank (oklass) == 1) {
+		if (m_class_is_array_special_interface (klass) && m_class_get_rank (oklass) == 1 && m_class_get_byval_arg (oklass)->type == MONO_TYPE_SZARRAY) {
```

Comment:
Won't this need JIT/interp changes as well ?

Reply:
As far as I can tell in both cases the calls end up in `mono_class_is_assignable_from_general` at runtime.

The interp seems to call `mono_interp_isinst` and the JIT/AOT ends up in `mono_marshal_isinst_with_cache` (which ends up `mono_object_isinst_vtable_mbyref`)